libnetwork/network/interface.go

```diff
@@ -53,146 +47,69 @@ func NetworkBackend(store storage.Store, conf *config.Config, syslog bool) (type
 		}
 	}
 
-	switch backend {
-	case types.Netavark:
-		netavarkBin, err := conf.FindHelperBinary(netavarkBinary, false)
-		if err != nil {
-			return "", nil, err
-		}
+	return backendFromType(backend, store, conf, syslog)
+}
 
-		aardvarkBin, _ := conf.FindHelperBinary(aardvarkBinary, false)
+func defaultNetworkBackend(store storage.Store, conf *config.Config) (backend types.NetworkBackend, err error) {
+	err = nil
 
-		confDir := conf.Network.NetworkConfigDir
-		if confDir == "" {
-			confDir = getDefaultNetavarkConfigDir(store)
-		}
+	file := filepath.Join(store.GraphRoot(), defaultNetworkBackendFileName)
 
-		// We cannot use the runroot for rootful since the network namespace is shared for all
-		// libpod instances they also have to share the same ipam db.
-		// For rootless we have our own network namespace per libpod instances,
-		// so this is not a problem there.
-		runDir := netavarkRunDir
-		if unshare.IsRootless() {
-			runDir = filepath.Join(store.RunRoot(), "networks")
-		}
+	createBackendSaveFunc := func(backendT types.NetworkBackend) func() {
+		return func() {
+			// only write when there is no error
+			if err == nil {
+				if err := ioutils.AtomicWriteFile(file, []byte(backendT), 0o644); err != nil {
+					logrus.Errorf("could not write network backend to file: %v", err)
+				}
+			}
 
-		netInt, err := netavark.NewNetworkInterface(&netavark.InitConfig{
-			Config:           conf,
-			NetworkConfigDir: confDir,
-			NetworkRunDir:    runDir,
-			NetavarkBinary:   netavarkBin,
-			AardvarkBinary:   aardvarkBin,
-			Syslog:           syslog,
-		})
-		return types.Netavark, netInt, err
-	case types.CNI:
-		netInt, err := getCniInterface(conf)
-		return types.CNI, netInt, err
-
-	default:
-		return "", nil, fmt.Errorf("unsupported network backend %q, check network_backend in containers.conf", backend)
+		}
 	}
-}
 
-func defaultNetworkBackend(store storage.Store, conf *config.Config) (backend types.NetworkBackend, err error) {
 	// read defaultNetworkBackend file
-	file := filepath.Join(store.GraphRoot(), defaultNetworkBackendFileName)
 	b, err := os.ReadFile(file)
 	if err == nil {
 		val := string(b)
+
+		// if the network backend has been already set previously,
+		// handle the values depending on whether CNI is supported and
+		// whether the network backend is explicitly configured
 		if val == string(types.Netavark) {
+			// netavark is always good
 			return types.Netavark, nil
+		} else if val == string(types.CNI) {
+			//
+			if cniSupported {
+
+				return types.CNI, nil
+			} else if conf.Network.NetworkBackend == "" {
```

Comment:
no need to check this, defaultNetworkBackend() should only be called when it is already empty so we know it must always be empty here